src/codemirror/extensions/autocomplete.ts

```diff
@@ -0,0 +1,180 @@
+} from "@codemirror/autocomplete";
+import { extractEditorStateInfo } from "../utils/extensions-utils";
+import { AutocompletionsMetadata } from "@/types";
+
+function isNewTagContext(lineText: string): boolean {
+  return /<\w*$/.test(lineText);
+}
+
+function generateComponentNameCompletions(
+  partialName: string,
+  options: AutocompletionsMetadata
+): Completion[] {
+  return Object.keys(options)
+    .filter((componentName) =>
+      componentName.toLowerCase().startsWith(partialName.toLowerCase())
+    )
+    .map((componentName) => ({
+      label: componentName,
+      section: "Components",
+      type: "keyword",
+      apply: (view, completion, from, to) => {
+        // insert the component name and a space
+        const insertText = `${completion.label} `;
+        view.dispatch({
+          changes: { from, to, insert: insertText },
+          // move the cursor to the end of the new inserted text
+          selection: { anchor: from + insertText.length },
+        });
+
+        // we need to trigger autocompletion to immediately show prop completions after inserting the component name
+        requestAnimationFrame(() => {
+          startCompletion(view);
+        });
+      },
+    }));
+}
+
+function extractAlreadyUsedProps(
+  fullLineText: string,
+  cursorPos: number
+): Set<string> {
+  const tagStart = fullLineText.lastIndexOf("<", cursorPos);
+  let tagEnd = fullLineText.indexOf(">", cursorPos);
+  if (tagEnd === -1) {
+    // tags that aren't closed yet
+    tagEnd = fullLineText.length;
+  }
+
+  const tagContent = fullLineText.substring(tagStart, tagEnd);
+
+  // props in the format propName="value", propName={value}, and just propName (flag-boolean)
+  const propRegex = /(\w+)(?=\s*=|\s*\/>|>|$)/g;
+  const usedProps = new Set<string>();
+  let match;
+
+  while ((match = propRegex.exec(tagContent))) {
+    usedProps.add(match[1]);
+  }
+
+  return usedProps;
+}
+
+function generatePropCompletions(
+  componentName: keyof AutocompletionsMetadata,
+  options: AutocompletionsMetadata,
+  partialPropName: string,
+  usedProps: Set<string>,
+  lineTextUpToCursor: string
+): Completion[] {
+  const componentProps = options[componentName];
+  if (!componentProps?.length) {
+    return [];
+  }
+
+  const normalizedProps = componentProps.flatMap((propItem) =>
+    Object.entries(propItem).map(([name, propInfo]) => ({ name, ...propInfo }))
+  );
+
+  return normalizedProps
+    .filter(
+      ({ name }) => !usedProps.has(name) && name.startsWith(partialPropName)
+    )
+    .map((item) => {
+      console.log({ item });
+      return item;
+    })
+    .map(({ name, type, required, defaultValue, description }) => ({
+      label: name,
+      boost: required ? 1 : 0,
+      detail: Array.isArray(type) ? type.join(" | ") : type,
+      info:
+        (description ? `${description}` : "") +
+        (description && defaultValue ? " | " : "") +
+        (defaultValue ? ` Defaults to: ${defaultValue}` : ""),
```

Comment:
A lot of string manipulation, is it safe?

Reply:
basically there shouldn't be an issue with it
the thing is that both description and defaultValue are not guaranteed to exist
So I want to show description and show separation `|` in case I have both defaultValue and description
See example in the gif above

Reply:
Haha ok, the gif is really low quality so it's hard to understand what's written